ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java

```diff
@@ -310,20 +310,23 @@ CompletableFuture<Long> getAppliedIndexFuture() {
       return appliedIndexFuture;
     }
 
-    boolean isApplied(LogEntryProto logEntry) {
+    boolean isFirstLogInCurrentTerm(LogEntryProto logEntry) {
       if (appliedIndexFuture.isDone()) {
-        return true;
+        return false;
       }
-      final long appliedIndex = logEntry != null? logEntry.getIndex(): server.getState().getLastAppliedIndex();
-      if (appliedIndex >= startIndex) {
+      final long appliedIndex = logEntry.getIndex();
+      if (appliedIndex == startIndex) {
         appliedIndexFuture.complete(appliedIndex);
-        LOG.info("leader is ready since appliedIndex == {} >= startIndex == {}",
-            appliedIndex, startIndex);
+        LOG.info("leader is ready since appliedIndex == startIndex == {}", startIndex);
         return true;
       } else {
         return false;
       }
     }
+
+    boolean isApplied() {
+      return appliedIndexFuture.isDone();
```

Comment:
Need to check cancelation and exception.
```java
      return JavaUtils.isCompletedNormally(appliedIndexFuture);
```
```java
//JavaUtils
  static boolean isCompletedNormally(CompletableFuture<?> future) {
    return future.isDone() && !future.isCancelled() && !future.isCompletedExceptionally();
  }
```

Reply:
Good idea!